sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/RocksDbInstance.scala

```diff
@@ -0,0 +1,434 @@
+      case Some(valueInBytes) =>
+        val value = new UnsafeRow(valueSchema.fields.length)
+        value.pointTo(valueInBytes, valueInBytes.length)
+        value
+      case None =>
+        null
+    }
+  }
+
+  override def commit(backupPath: Option[String] = None): Unit = {
+    verify(txn != null, "Start Transaction before fetching any key-value")
+    // printTrxStats
+    try {
+      val file = new File(dbPath, identifier.toUpperCase(Locale.ROOT))
+      file.createNewFile()
+      txn.commit()
+      txn.close()
+      txn = null
+      backupPath.foreach(f => createCheckpoint(otdb.asInstanceOf[RocksDB], f))
+    } catch {
+      case e: Exception =>
+        log.error(s"Unable to commit the transactions. Error message = ${e.getMessage}")
+        throw e
+    }
+  }
+
+  def printTrxStats(): Unit = {
+    verify(txn != null, "No open Transaction")
+    logInfo(s"""
+         | deletes = ${txn.getNumDeletes}
+         | numKeys = ${txn.getNumKeys}
+         | puts =  ${txn.getNumPuts}
+         | time =  ${txn.getElapsedTime}
+       """.stripMargin)
+  }
+
+  override def abort(): Unit = {
+    verify(txn != null, "No Transaction to abort")
+    txn.rollbackToSavePoint()
+    txn.close()
+    txn = null
+  }
+
+  override def close(): Unit = {
+    verify(isOpen(), "No DB to close")
+    readOptions.close()
+    writeOptions.close()
+    logDebug("Closing the transaction db")
+    otdb.close()
+    otdb = null
+  }
+
+  override def iterator(closeDbOnCompletion: Boolean): Iterator[UnsafeRowPair] = {
+    verify(txn != null, "Transaction is not set")
+    verify(
+      closeDbOnCompletion == false,
+      "Cannot close a DB without aborting/commiting the transactions")
+    val readOptions = new ReadOptions()
+    val itr: RocksIterator = txn.getIterator(readOptions)
+    Option(itr) match {
+      case Some(i) =>
+        logDebug(s"creating iterator from transaction DB")
+        createUnsafeRowPairIterator(i, readOptions, false)
+      case None =>
+        Iterator.empty
+    }
+  }
+
+}
+
+object RocksDbInstance {
+
+  RocksDB.loadLibrary()
+
+  private val destroyOptions: Options = new Options()
+
+  val lRUCache = new LRUCache(1024 * 1024 * 1024, 6, false, 0.05)
+
+  def destroyDB(path: String): Unit = {
+    val f: File = new File(path)
+    if (f.exists()) {
+      RocksDB.destroyDB(path, destroyOptions)
+      FileUtils.deleteDirectory(f)
+    }
+  }
+
+  def restoreFromBackup(backupDir: String, dbDir: String): Unit = {
+    val (result, elapsedMs) = Utils.timeTakenMs {
+      val backupableDBOptions = new BackupableDBOptions(backupDir)
+      val be = BackupEngine.open(Env.getDefault, backupableDBOptions)
+      val restoreOptions = new RestoreOptions(false)
+      be.restoreDbFromLatestBackup(dbDir, dbDir, restoreOptions)
+      restoreOptions.close()
+      backupableDBOptions.close()
+      be.close()
+    }
+  }
+
+  def verify(condition: => Boolean, msg: String): Unit = {
```

Comment:
Do we need to reimplement Scala `require`?

Reply:
Fixed